### ISSUE TITLE
OneNote Integration with Drupal/WordPress

### ISSUE THREAD

**Issue by tadhg-ohiggins** [18F/dos-ops-center#263]
- [ ] Get access to Ops Center OneNote API.
- [ ] Determine that the API allows for the extraction of notes in such a way that would allow them to be exported to a KMS.
- [ ] Stand up cloud.gov instances of WordPress and Drupal in the `state-prototyping` space.
- [ ] Successfully export basic OneNote content to those instances (this will not be a perfect export, but accomplishing this will establish what some of the problems are).

**Comment by karareinsel** [comment]
Waiting on ExecTech for the OneNote API key.

**Comment by beebouttw** [comment]
Forwarding from Tom Lawrence of ExecTech:

It appears the that Microsoft has a huge library of REST APIs for many of the Office365 products. That excites for a few more taskers you have asked of us…back on track.
The OneNote root URL is https://graph.microsoft.com/{version}/{location}/onenote/ . And per this article version 1.0 is stable production code. 
 
OneNote specific List of APIs are: 
https://docs.microsoft.com/en-us/graph/api/resources/onenote?view=graph-rest-1.0
https://docs.microsoft.com/en-us/graph/api/onenote-list-notebooks?view=graph-rest-1.0&tabs=cs
https://docs.microsoft.com/en-us/graph/api/onenote-post-notebooks?view=graph-rest-1.0&tabs=cs
https://docs.microsoft.com/en-us/graph/api/onenote-list-sections?view=graph-rest-1.0&tabs=cs
https://docs.microsoft.com/en-us/graph/api/onenote-list-sectiongroups?view=graph-rest-1.0&tabs=cs
https://docs.microsoft.com/en-us/graph/api/onenote-list-pages?view=graph-rest-1.0
https://docs.microsoft.com/en-us/graph/api/onenote-post-pages?view=graph-rest-1.0

In short, the APIs can view the section or entire notebook. It can also create pages, new sections and entire new notebooks. 

So nice, but we are the government, that means permission to do so. Well the good news is there is a separate section on permissions as well. This is also required if a solution wants to access your specific Notebook. 
https://docs.microsoft.com/en-us/graph/permissions-reference#notes-permissions

The architecture looks like this. 

 

So what is that SDK? Well, if you have Azure, and we do, you can write code on the follow platforms that can interact with Azure. Note, Office365 sits on top of Azure. Hmmmm. 

SDKs are currently available for the following languages and platforms:
•	Android
•	Angular
•	ASP.NET
•	iOS
•	Javascript
•	Node.js
•	Java
•	PHP
•	Python
•	Ruby
Office 365 developer subscription
When building applications using Microsoft Graph, we recommend that you get a free Office 365 developer subscription by signing up for the Office 365 Developer Program.

**Comment by beebouttw** [comment]
The latest from Tom, with my response below:

Anything is possible. With our IRM and DS colleagues it’s the details that matter. 

Data can be pushed out of OpenNet. It’s a little more than semantics, but it’s not an obstacle. We have formally told IRM that for Pangea to work, inter-agency clearances are a must have. 

For early Development we can do a few things, but the key is to understand that 18F dev can be view in the same light as any contractor of State. So don’t assume they can’t develop on OpenNet. That’s a main point of changing to DevOps. Jason and I are currently working with IRM on Azure and AWS instants for our efforts. If we know how 18F wants to approach, it is still early enough to add this to our negotiation with IRM.

Dummy data is only required when your system will be used for controlling SBU such as PII, HEPA and other content that has statues or laws protecting. Most of the products you produce on OpenNet are operational in nature. Meaning SBU, but shareable. If 18F plans to Dev offline and bring code into testing, it’s on them to get dummy data. One example we did for our briefing book dev, we used Canterbury Tails. 

Sorry to answer in a more generic response. The point is we have many options, but understanding how 18F plans to execute matters. if they are assuming on their end and we can help in that plan of execution, please let us know.
-tom

-------------------------------------------------------

Thanks, Tom, that makes sense.  There are a lot of encouraging statements in there vis a vis our ongoing efforts.  It sounds like the best way forward with this for the short term might be avoiding OpenNet altogether.  

My understanding of what 18F is trying to do is to look at the data coming out of a OneNote API and see how well the formatting and features hand over to a few external programs, notably Wordpress and Drupal out of the box, as well as a few other text editors we’ve been looking at.  I’m somewhat putting words in their mouth here, but looking at the collaborative editing testing we’ve already done, this seems like the standard MO.  Since the data itself needn’t be sensitive (it could also be the Canterbury Tales for consistency’s sake), the real goal here is to do a little hands-on testing with the OneNote API.  Long term integration is certainly a dream, assuming the short-term testing goes well.

Do you anticipate having the same restrictions if we could get a DIN machine running OneNote filled with dummy data?  That might provide useful feedback on how well the API serves our purposes without having to forge an entire trail with IRM just yet.

Thanks,
Tyler
